app/components/CredentialCard.tsx

```diff
@@ -22,39 +22,63 @@ import TimeAgo from './TimeAgo'
 import { ToolbarButtonGroup } from './ToolbarButtonGroup'
 import type { RedactedClientCredential } from '~/routes/user.credentials/client_credentials.server'
 
+const WARNING_DAYS = 21 * 24 * 60 * 60 * 1000 // 21 days in milliseconds
+
 export default function CredentialCard({
   name,
   client_id,
   created,
   scope,
   scopeDescription,
   lastUsed,
+  expired,
 }: RedactedClientCredential & { scopeDescription?: string }) {
   const ref = useRef<ModalRef>(null)
   const fetcher = useFetcher()
   const [searchParams] = useSearchParams()
   const disabled = fetcher.state !== 'idle'
   const alerts = searchParams.getAll('alerts')
+  const showExpirationWarning =
+    (lastUsed === undefined || Date.now() - lastUsed > WARNING_DAYS) && !expired
```

Comment:
This version is correct for undefined, NaN, null, etc.

```suggestion
  const showExpirationWarning =
    !(expired || Date.now() - lastUsed <= WARNING_DAYS)
```